hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -60,7 +60,7 @@ object InferPType {
       case Void() => PVoid
       case Cast(ir, t) => {
         InferPType(ir, env)
-        PType.canonical(t, ir.pType2.required)
+        PType.canonical(t, ir._pType2.required)
```

Comment:
all pType2 should be the protected member _pType2 (which avoids the null branch check).

Reply:
To make the review easier, I note the 2 cases of bugs in master InferPType, and then the beginning of the new stuff.

Reply:
why this change? It won't matter for performance and makes the code a little harder to read. I think the non-null assertion isn't a bad thing to have here

Reply:
can we at least punt to a different PR? makes the diff much more complicated

Reply:
Sure. I think just because the public accessor wasn't meant to be used inside the inference pass. Not sure why it proliferated.

Reply:
done